src/arithmetic/arithmetic_expression.c

```diff
@@ -477,9 +479,10 @@ SIValue AR_SUBSTRING(SIValue *argv, int argc) {
 
     assert(start >=0 && length >=0);
     
-    substring = calloc(length+1, sizeof(char));
+    substring = malloc(length+1 * sizeof(char));
     memcpy(substring, original + start, length);
-    return SI_StringValC(substring);
+    substring[length] = '\0';
+    return SI_StringVal(substring);
```

Comment:
SI_StringVal duplicates, need to free substring (would be better if this was allocated on the stack).